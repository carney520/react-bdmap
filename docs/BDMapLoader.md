baidu map 是通过`script`异步加载的, 应该使用`BDMapLoader`作为`BDMap`的上级，`BDMapLoader`会在加载成功后才会渲染子节点，
所以最佳实践是将`BDMap`放在单独的组件中， 这样可以确保可以安全地访问`window`下的`BMap`属性。如:

```jsx static
import React from 'react'
import BDMapLoader from 'react-bdmap/BDMapLoader'
import BDMap from 'react-bdmap/BDMap'

// 可以安全地在所有'实例属性'、'实例方法'中访问BMap
class MyMap extends React.Component {
  state = {
    center: new BMap.Point(123, 78),
  }

  render() {
    return <BMap center={this.state.center}>{/* controls, overlays */}</BMap>
  }
}

class MyApp {
  render() {
    return (
      <BDMapLoader
        apiKey="MY_KEY"
        fallback={err => {
          return err ? 'Failed to load' : 'Loading...'
        }}
      >
        <MyMap />
      </BDMapLoader>
    )
  }
}
```

使用 fallback 来自定义展示加载状态

```jsx
function App() {
  return (
    <BDMapLoader
      apiKey="1XWjAMBIusA3EL4G61lXS0AliZd0l7bF"
      fallback={error => (error ? `Failed to load: ${error}` : 'loading')}
    >
      <div>loaded</div>
    </BDMapLoader>
  )
}

class Example extends React.Component {
  constructor() {
    super()
    this.state = {
      show: false,
    }
  }

  render() {
    return (
      <div>
        {this.state.show && <App />}
        <div>
          <button onClick={() => this.setState({ show: true })}>show</button>
        </div>
      </div>
    )
  }
}

;<Example />
```
