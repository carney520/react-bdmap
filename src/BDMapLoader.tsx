/**
 * 初始化和加载baidu地图
 */
import React from 'react'
import { delay, importScript } from './utils'

declare global {
  interface Window {
    loadBDMap?: () => void
    BMap: typeof BMap
  }
}

export interface BDMapLoaderProps {
  // api key
  apiKey: string
  // loading or load failed
  fallback?: (error?: Error) => React.ReactNode
}

interface State {
  loaded: boolean
  error?: Error
}

const DEFAULT_RETRY_TIME = 3

export default class BDMapLoader extends React.Component<BDMapLoaderProps> {
  public state: State = {
    loaded: !!window.BMap,
  }

  public constructor(props: BDMapLoaderProps) {
    super(props)
    if (this.props.apiKey == null) {
      throw new TypeError('BDMap: apiKey is required')
    }
  }

  public componentDidMount() {
    if (window.BMap == null) {
      this.loadMap()
      return
    }
  }

  public render() {
    return this.state.loaded ? (
      this.props.children
    ) : this.props.fallback ? (
      this.props.fallback(this.state.error)
    ) : this.state.error ? (
      <div style={{ color: 'red' }}>{this.state.error.message}</div>
    ) : null
  }

  /**
   * load bdmap in script tag
   */
  private async loadMap() {
    const src = `//api.map.baidu.com/api?v=3.0&ak=${this.props.apiKey}&callback=loadBDMap`
    window.loadBDMap = this.finish
    for (let i = 0; i < DEFAULT_RETRY_TIME; i++) {
      try {
        await importScript(src)
        break
      } catch (error) {
        if (i === DEFAULT_RETRY_TIME - 1) {
          // 加载失败提示
          this.setState({ error: new Error(`Failed to load Baidu Map: ${error.message}`) })
        }
        await delay(i * 1000)
      }
    }
  }

  private finish = () => {
    this.setState({
      loaded: true,
    })
  }
}
