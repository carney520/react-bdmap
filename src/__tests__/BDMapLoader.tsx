import React from 'react'
import BDMapLoader from '../BDMapLoader'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'

test('throw error when apiKey not present', () => {
  const render = () => {
    // @ts-ignore
    shallow(<BDMapLoader />)
  }
  expect(render).toThrow()
})

describe('BMap loaded', () => {
  beforeAll(() => {
    // @ts-ignore
    window.BMap = {}
  })

  afterAll(() => {
    // @ts-ignore
    window.BMap = undefined
  })

  it('should match snapshot', () => {
    const comp = renderer
      .create(
        <BDMapLoader apiKey="test">
          <div>map</div>
        </BDMapLoader>,
      )
      .toJSON()
    expect(comp).toMatchSnapshot()
  })

  test('render children when BMap loaded', () => {
    const component = shallow(
      <BDMapLoader apiKey="test">
        <div>map</div>
      </BDMapLoader>,
    )
    expect(component.containsMatchingElement(<div>map</div>)).toBeTruthy()
  })
})
