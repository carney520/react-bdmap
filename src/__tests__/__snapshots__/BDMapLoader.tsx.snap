// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BMap loaded should match snapshot 1`] = `
<div>
  map
</div>
`;
